packages/gatsby/cache-dir/app.js

```diff
@@ -57,7 +57,7 @@ apiRunnerAsync(`onClientEntry`).then(() => {
   loader.addDevRequires(syncRequires)
 
   loader.getResourcesForPathname(window.location.pathname).then(() => {
-    let Root = hot(module)(preferDefault(require(`./root`)))
+    let Root = preferDefault(require(`./root`))
```

Comment:
New `/root/hot` API will not stand asynchronously between the first part and the second.

Reply:
Do you mean that `hot` should stay in this module? i.e.
```js
let Root = hot(preferDefault(require(`./root`)))
```

That was the reason for the original `hot update was not successful` error

Reply:
`hot` or `import 'react-hot-loader/root` are calling `module.hot.accept` for the current module, thus setting self-acceptance.
It's expected that some action would take a place on module hot update ____immediately____ and in a synchronous mode. Using `hot` inside indirectly called function are breaking some expectations - "module" could not accept it, as long as the "real action" is not yet called.

Here - if you want to make `./root` hot - you have to make IT hot. And that's was done.

The second moment is about HMR and an update propagation.
When you change a file webpack _bubbles_ update to parent, unless parent will accept the change, then webpack will update parent and everything below it.
We had WFT-level issues if "parent" is the module with `react-dom/render` - everything got wiped and regenerated, making reconciliation impossible (`Symbols` could be updated for example). 

This is not quite clear from RHL documentation, sorry. The difference is between `self-accept` in v4, and `child-accept`(ie _module.hot.accept('./containers/App'_) in v3.

Here you might want to accept only `./root`, but would accept _everything_.